### PR TITLE
In non strict mode, allow `let` as an identifier in expression position.

### LINKED ISSUE #1289: allow `let` as an identifier in expression position in sloppy mode
Fails to parse:

```js
for (let a = let;;);
```

```js
for(let {a: a = let};;);
```

```js
for(let [a = let];;);
```

```js
for(var a in let);
```

```js
for(var a of let);
```

```js
(let[a]=0)
```

```js
let.let=0
```

```js
let
```

### PR DESCRIPTION
**Pull request by ariya**
Fixes #1289

### CODE REVIEW

**Comment by mikesherov** [comment]
Looking good. Perhaps add a test for use in implicit strict mode inside modules?

File: esprima.js [hunk 1]
```diff
@@ -3955,6 +3956,12 @@
             }
         }
 
+        if (matchKeyword('let') || matchContextualKeyword('let')) {
```
Comment (mikesherov) [inline esprima.js]: Which test causes `matchKeyword('let')` to return true here?

Comment (mikesherov) [inline esprima.js]: Consider adding a function called `matchLet` as this logic is used in several places.

File: esprima.js [hunk 2]
```diff
@@ -4095,6 +4100,56 @@
         return list;
     }
 
+
+    function tokenizerState() {
```
Comment (mikesherov) [inline esprima.js]: Admitting that I'm still relatively green when it comes to understanding every PR, it seems fishy that this change would be the first to require a double lookahead and restore.

Comment (michaelficarra) [inline esprima.js]: It's necessary. See the two token lookahead in [section 13.5](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-expression-statement) and the associated NOTE text.

Comment (mikesherov) [inline esprima.js]: TIL

**Comment by michaelficarra** [comment]
I would leave `let` as a Keyword. It makes it easier to stop accepting it outside of strict mode.

File: esprima.js (hunk 1 above)
Comment (ariya) [inline esprima.js]: > Which test causes matchKeyword('let') to return true here?

In a case like this `esprima.parse('let x', { sourceType: 'module'})`. See function `advance()`, it automatically upgrades any strict mode reserved word (including `let`) to Keyword.

Comment (ariya) [inline esprima.js]: > Consider adding a function called matchLet as this logic is used in several places.

I thought about this, but I think it's only used twice and adding an extra mental note (assuming you're already familiar with `matchKeyword` and `matchContextualKeyword`) does not seem to be worth the hassle.

**Comment by ariya** [comment]
> Perhaps add a test for use in implicit strict mode inside modules?

That should be covered by `test/fixtures/ES6/lexical-declaration/module_let.js`.

**Comment by mikesherov** [comment]
OK, LGTM.

**Comment by mikesherov** [comment]
Might want to address @michaelficarra's concerns though.

**Comment by ariya** [comment]
> I would leave let as a Keyword. It makes it easier to stop accepting it outside of strict mode.

I originally tried this and it was too cumbersome (for some forgotten reasons). Now that I understand the situation a bit better, let me look at this again.

**Comment by ariya** [comment]
BTW that Circle CI failed build is a false alarm (the downstream check fails on JSCS, again).

**Comment by michaelficarra** [comment]
JSCS!!!! :confounded: :weary:

**Comment by ariya** [comment]
I found my previous attempt where `let` was still a keyword. Comparing both versions, I think the keyword should be used for now. I have no strong preference of one or another, but reducing the drastic change for the consumer is always a (minor) bonus.

I'll update this pull request shortly.

**Comment by ariya** [comment]
All tests, including the downstream project tests, now pass with flying colors!

File: esprima.js [hunk 3]
```diff
@@ -2746,7 +2741,12 @@
             return parseArrayPattern(params, kind);
         } else if (match('{')) {
             return parseObjectPattern(params, kind);
+        } else if (matchKeyword('let')) {
+            if (kind === 'const' || kind === 'let') {
```
Comment (mikesherov) [inline esprima.js]: Could this instead be `!== 'var'` and be folded into previous else if?

Comment (mikesherov) [inline esprima.js]: Never mind.

**Pull request closed without merging**